### DISCUSSION TITLE
WebSocket Protocol Support

### DISCUSSION THREAD

**Discussion by isuru89** [mockoon/mockoon#1209] [Show and tell]
Hi All,

I have done a successful PoC on WebSocket support on Mockoon. 
Link to work-in-progress branch: [here](https://github.com/mockoon/mockoon/compare/main...isuru89:mockoon:poc/ws)

Expectation is to start simple functionalities and make it improve further. My implementation has below features and limitations:

**Features**:
  - Can define WebSocket protocol end points and consume them. (e.g. `ws://localhost:3000/<sub-path>`)
  - Can mix both WebSocket and HTTP endpoints together in the same environment.
  - Ability to respond based on the content of incoming message (full-duplex conversations)
  - Ability to broadcast same data to all the connected clients of a particular WebSocket endpoint (broadcast mode)
  - Ability to unicast streaming data for each connected client of a particular WebSocket endpoint (one-to-one mode)

**Limitations**:
  - No Socket.IO support. Only `ws/wss` portocols.
  - Path parameters are not supported for WebSockets (yet)
  - WebSockets does not show in logs. Nor recording supported.
  - Cannot forward and then broadcast a message from another route to a WebSocket end point.
  - No multicast support (this feature is in socket.io with rooms). Only one-to-one and broadcast streaming is supported.
  - No authentications assumed.

**Here are some screenshots.**

 * Adding a new WebSocket from routes menu
![mockoon-add-ws](https://github.com/mockoon/mockoon/assets/1499404/e66de991-56b6-4f2a-b839-c1e203f8ea40)

 * View of a websocket route with newly introduced options.
![mockoon-ws-route-view](https://github.com/mockoon/mockoon/assets/1499404/f580865e-6076-4c96-a2ae-41aa889a7af4)

**Comment by 255kb** [comment]
Thanks @isuru89 for this awesome work. I can't wait to put my hands on your PR!
